### PR TITLE
Added optional `--sat_per_byte` arg to openchannel for lnd

### LINKED ISSUE #1829: Add optional `--sat_per_byte` arg to openchannel
for btc/ltc only, others should throw an invalid arg error

### PR DESCRIPTION
**Pull request by rsercano**
resolves #1829

### CODE REVIEW

**Comment by raladev** [comment]
- [ ] we have --fee parameter for withdrawwallet command, so it would be good to have same fee naming for all commands.

[Screenshot from 2020-08-27 14-16-27](https://user-images.githubusercontent.com/29906866/91438309-cd366a80-e873-11ea-9541-a2e3f222cd4f.png)

- [ ] Should we add same fee filed for closechannel call? @kilrau @rsercano 

- [ ] if user set too small fee rate channel will be opened with default fee rate. Should we print about that and prevent channel opening in this case? @kilrau @rsercano 
```
2020-08-27 11:28:05.606 [INF] SWPR: Manual fee rate input of 250 sat/kw is too low, using 253 sat/kw instead
2020-08-27 11:28:05.606 [DBG] RPCS: [openchannel]: using fee of 253 sat/kw for funding tx
```
- Also i tried different values (1, 250, 300, 13000) and i saw difference in lndbtc logs 
```
2020-08-27 11:34:49.462 [DBG] RPCS: [openchannel] request to NodeKey() allocation(us=10000000, them=0)
2020-08-27 11:34:49.462 [DBG] RPCS: [openchannel]: using fee of 75000 sat/kw for funding tx
2020-08-27 11:34:49.462 [INF] FNDG: Initiating fundingRequest(local_amt=0.1 BTC (subtract_fees=false), push_amt=0 mSAT, chain_hash=683e86bd5c6d110d91b94b97137ba6bfe02dbbdb8e3dff722a669b5d69d77af6, peer=0252f366111259996a101c5a82c880e9c95c8278d7e668c40163c0db656161cbbf, dust_limit=0.00000573 BTC, min_confs=1)
2020-08-27 11:34:49.462 [INF] CHFD: Performing funding tx coin selection using 75000 sat/kw as fee rate
2020-08-27 11:34:49.475 [INF] FNDG: Target commit tx sat/kw for pendingID(60b7ea115f69784df02778ab26a58e3b851c7c123798e1b0e8a85c03d1f14d4b): 12500



2020-08-27 11:37:17.980 [INF] FNDG: Initiating fundingRequest(local_amt=0.1 BTC (subtract_fees=false), push_amt=0 mSAT, chain_hash=683e86bd5c6d110d91b94b97137ba6bfe02dbbdb8e3dff722a669b5d69d77af6, peer=0252f366111259996a101c5a82c880e9c95c8278d7e668c40163c0db656161cbbf, dust_limit=0.00000573 BTC, min_confs=1)
2020-08-27 11:37:17.980 [INF] CHFD: Performing funding tx coin selection using 3250000 sat/kw as fee rate
2020-08-27 11:37:17.990 [INF] FNDG: Target commit tx sat/kw for pendingID(d787cbc05bff6b32db12261886177946d0f359634b6e0f476c8534a12324c843): 12500
```
but i dont see any difference in` "commit_fee",  "commit_weight": "600",,"fee_per_kw": "12500"` parameters of opened channels:
```
{
    "channels": [
        {
            "active": true,
            "remote_pubkey": "0252f366111259996a101c5a82c880e9c95c8278d7e668c40163c0db656161cbbf",
            "channel_point": "6702394f5df7acacd783fb14834e1d2d75841e9a5177defcc2d97ed7fe8e863c:0",
            "chan_id": "97759777848885248",
            "capacity": "10000000",
            "local_balance": "9990950",
            "remote_balance": "0",
            "commit_fee": "9050",
            "commit_weight": "600",
            "fee_per_kw": "12500",
            "unsettled_balance": "0",
            "total_satoshis_sent": "0",
            "total_satoshis_received": "0",
            "num_updates": "0",
            "pending_htlcs": [
            ],
            "csv_delay": 1201,
            "private": false,
            "initiator": true,
            "chan_status_flags": "ChanStatusDefault",
            "local_chan_reserve_sat": "100000",
            "remote_chan_reserve_sat": "100000",
            "static_remote_key": true,
            "commitment_type": "STATIC_REMOTE_KEY",
            "lifetime": "475",
            "uptime": "475",
            "close_address": "",
            "push_amount_sat": "0",
            "thaw_height": 0,
            "local_constraints": {
                "csv_delay": 1201,
                "chan_reserve_sat": "100000",
                "dust_limit_sat": "573",
                "max_pending_amt_msat": "9900000000",
                "min_htlc_msat": "1",
                "max_accepted_htlcs": 483
            },
            "remote_constraints": {
                "csv_delay": 1201,
                "chan_reserve_sat": "100000",
                "dust_limit_sat": "573",
                "max_pending_amt_msat": "9900000000",
                "min_htlc_msat": "1",
                "max_accepted_htlcs": 483
            }
        },
        {
            "active": true,
            "remote_pubkey": "0252f366111259996a101c5a82c880e9c95c8278d7e668c40163c0db656161cbbf",
            "channel_point": "22accc1a6160578be795f5078e30eae0d400d478824c8c94935458282e5a0593:0",
            "chan_id": "97756479314001920",
            "capacity": "10000000",
            "local_balance": "9990950",
            "remote_balance": "0",
            "commit_fee": "9050",
            "commit_weight": "600",
            "fee_per_kw": "12500",
            "unsettled_balance": "0",
            "total_satoshis_sent": "0",
            "total_satoshis_received": "0",
            "num_updates": "0",
            "pending_htlcs": [
            ],
            "csv_delay": 1201,
            "private": false,
            "initiator": true,
            "chan_status_flags": "ChanStatusDefault",
            "local_chan_reserve_sat": "100000",
            "remote_chan_reserve_sat": "100000",
            "static_remote_key": true,
            "commitment_type": "STATIC_REMOTE_KEY",
            "lifetime": "656",
            "uptime": "656",
            "close_address": "",
            "push_amount_sat": "0",
            "thaw_height": 0,
            "local_constraints": {
                "csv_delay": 1201,
                "chan_reserve_sat": "100000",
                "dust_limit_sat": "573",
                "max_pending_amt_msat": "9900000000",
                "min_htlc_msat": "1",
                "max_accepted_htlcs": 483
            },
            "remote_constraints": {
                "csv_delay": 1201,
                "chan_reserve_sat": "100000",
                "dust_limit_sat": "573",
                "max_pending_amt_msat": "9900000000",
                "min_htlc_msat": "1",
                "max_accepted_htlcs": 483
            }
        },
        {
            "active": true,
            "remote_pubkey": "0252f366111259996a101c5a82c880e9c95c8278d7e668c40163c0db656161cbbf",
            "channel_point": "bdf4b702c293563dd341fb3f37d9b49bb583249b55ef1d7d3e912be32f596bc7:0",
            "chan_id": "97753180779118592",
            "capacity": "10000000",
            "local_balance": "9990950",
            "remote_balance": "0",
            "commit_fee": "9050",
            "commit_weight": "600",
            "fee_per_kw": "12500",
            "unsettled_balance": "0",
            "total_satoshis_sent": "0",
            "total_satoshis_received": "0",
            "num_updates": "0",
            "pending_htlcs": [
            ],
            "csv_delay": 1201,
            "private": false,
            "initiator": true,
            "chan_status_flags": "ChanStatusDefault",
            "local_chan_reserve_sat": "100000",
            "remote_chan_reserve_sat": "100000",
            "static_remote_key": true,
            "commitment_type": "STATIC_REMOTE_KEY",
            "lifetime": "836",
            "uptime": "836",
            "close_address": "",
            "push_amount_sat": "0",
            "thaw_height": 0,
            "local_constraints": {
                "csv_delay": 1201,
                "chan_reserve_sat": "100000",
                "dust_limit_sat": "573",
                "max_pending_amt_msat": "9900000000",
                "min_htlc_msat": "1",
                "max_accepted_htlcs": 483
            },
            "remote_constraints": {
                "csv_delay": 1201,
                "chan_reserve_sat": "100000",
                "dust_limit_sat": "573",
                "max_pending_amt_msat": "9900000000",
                "min_htlc_msat": "1",
                "max_accepted_htlcs": 483
            }
        },
        {
            "active": true,
            "remote_pubkey": "0252f366111259996a101c5a82c880e9c95c8278d7e668c40163c0db656161cbbf",
            "channel_point": "aae7765da6a687bb30b8feda407d1907914334a8d6a5c331c59607e9a39603cf:0",
            "chan_id": "97749882244235264",
            "capacity": "10000000",
            "local_balance": "9990950",
            "remote_balance": "0",
            "commit_fee": "9050",
            "commit_weight": "600",
            "fee_per_kw": "12500",
            "unsettled_balance": "0",
            "total_satoshis_sent": "0",
            "total_satoshis_received": "0",
            "num_updates": "0",
            "pending_htlcs": [
            ],
            "csv_delay": 1201,
            "private": false,
            "initiator": true,
            "chan_status_flags": "ChanStatusDefault",
            "local_chan_reserve_sat": "100000",
            "remote_chan_reserve_sat": "100000",
            "static_remote_key": true,
            "commitment_type": "STATIC_REMOTE_KEY",
            "lifetime": "1017",
            "uptime": "1017",
            "close_address": "",
            "push_amount_sat": "0",
            "thaw_height": 0,
            "local_constraints": {
                "csv_delay": 1201,
                "chan_reserve_sat": "100000",
                "dust_limit_sat": "573",
                "max_pending_amt_msat": "9900000000",
                "min_htlc_msat": "1",
                "max_accepted_htlcs": 483
            },
            "remote_constraints": {
                "csv_delay": 1201,
                "chan_reserve_sat": "100000",
                "dust_limit_sat": "573",
                "max_pending_amt_msat": "9900000000",
                "min_htlc_msat": "1",
                "max_accepted_htlcs": 483
            }
        }
    ]
}
```
How can i check that my fee rate was used?

**Review by raladev — CHANGES_REQUESTED** [review]
above

**Comment by kilrau** [comment]
>  we have --fee parameter for withdrawwallet command, so it would be good to have same fee naming for all commands.

:+1: 

 > Should we add same fee filed for closechannel call?

:+1: 

> if user set too small fee rate channel will be opened with default fee rate. Should we print about that and prevent channel opening in this case?

No need for now I think.

**Comment by michael1011** [comment]
> but i did not see any difference in "commit_fee", "commit_weight": "600",,"fee_per_kw": "12500" parameters of opened channels:

Those values will remain unchanged. The flag is only for the fee of the funding transaction.

> How can i check that my fee rate was used?

With a block explorer in case you are on testnet or mainnet. Else you would have to query the funding transaction and all of the transactions of its inputs and calculate the fee yourself.

**Comment by rsercano** [comment]
Let me know if there's anything left to do on this PR

**Comment by kilrau** [comment]
> Let me know if there's anything left to do on this PR

Yes, these two: https://github.com/ExchangeUnion/xud/pull/1832#issuecomment-681945674

**Comment by rsercano** [comment]
Do we want to apply these changes within this PR, or seperate it with another issues?

**Comment by raladev** [comment]
> I mean do we want to apply these changes within this PR, or seperate it with another issues?

in this one.

**Comment by kilrau** [comment]
> > I mean do we want to apply these changes within this PR, or seperate it with another issues?
> 
> in this one.

Agreed

**Comment by rsercano** [comment]
Ah sorry my misunderstanding, fixing now.

File: lib/Config.ts [hunk 1]
```diff
@@ -358,7 +358,7 @@ class Config {
   }
 
   private getDefaultLogLevel = (): string => {
-    return process.env.NODE_ENV === 'production' ? Level.Info : Level.Debug;
+    return process.env.NODE_ENV === 'production' ? Level.Info : Level.Trace;
```
Comment (raladev) [inline lib/Config.ts:361]: looks like unnecessary change )

**Review by raladev — CHANGES_REQUESTED** [review]

File: lib/Config.ts (hunk 1 above)
Comment (rsercano) [reply lib/Config.ts]: Sorry my bad

**Review by raladev — DISMISSED** [review]

**Comment by rsercano** [comment]
Resolved conflict

**Review by raladev — DISMISSED** [review]

File: lib/cli/commands/closechannel.ts [hunk 2]
```diff
@@ -28,9 +28,15 @@ export const builder = (argv: Argv) => argv
...
   })
+  .option('fee', {
+    type: 'number',
+    description: 'the manual fee rate set in sat/byte that should be used when crafting the closure transaction in the channel',
+    default: 0,
+  })
   .example('$0 closechannel BTC 028599d05b18c0c3f8028915a17d603416f7276c822b6b2d20e71a3502bd0f9e0b', 'close BTC channels by node key')
   .example('$0 closechannel BTC CheeseMonkey', 'close BTC channels by alias')
   .example('$0 closechannel BTC CheeseMonkey --force', 'force close BTC channels by alias')
+  .example('$0 closechannel BTC CheeseMonkey 25', 'close BTC channels by alias with 25 sat/byte fee')
```
Comment (sangaman) [inline lib/cli/commands/closechannel.ts:39]: This command doesn't work, because the order of the args the 25 is assigned to the `force` variable. I'd make this command `$0 closechannel BTC CheeseMonkey --fee 25`.

File: lib/lndclient/LndClient.ts [hunk 3]
```diff
@@ -1083,15 +1085,15 @@ class LndClient extends SwapClient {
...
+        const closePromise = this.closeChannelSync(fundingTxId, Number(outputIndex), force, fee);
         closePromises.push(closePromise);
       }
     });
     await Promise.all(closePromises);
   }
 
   /** A synchronous helper method for the closeChannel call */
-  public closeChannelSync = (fundingTxId: string, outputIndex: number, force: boolean): Promise<void> => {
+  public closeChannelSync = (fundingTxId: string, outputIndex: number, force: boolean, fee: number): Promise<void> => {
```
Comment (sangaman) [inline lib/lndclient/LndClient.ts:1096]: Minor but I think `fee` can/should still be optional here by just assigning `fee = 0`.

**Review by sangaman — DISMISSED** [review]
Just a couple small things but otherwise looks good.

File: lib/cli/commands/closechannel.ts (hunk 2 above)
Comment (rsercano) [reply lib/cli/commands/closechannel.ts:39]: Ah true that

**Review by raladev — DISMISSED** [review]

**Comment by sangaman** [comment]
@rsercano needs a rebase.

**Review by raladev — DISMISSED** [review]

**Comment by rsercano** [comment]
done @sangaman

**Review by raladev — APPROVED** [review]

**Comment by rsercano** [comment]
Rebased once more, would you please recheck and squash merge @raladev @sangaman

**Review by raladev — APPROVED** [review]

**Comment by sangaman** [comment]
I'm going to tidy up the commit history/message and merge.

**Comment by kilrau** [comment]
I think you forgot to press this button ;) - Still needs an approving review to be merged.

> I'm going to tidy up the commit history/message and merge.

Ok, but why not use the `Squash and merge` button here in the GitHub UI (also allows to clean up the commit message)? Like this one can't forget merge since it does it all at once.

**Review by raladev — APPROVED** [review]
